lib/srv/server/ssm_install.go

```diff
@@ -104,6 +147,119 @@ func (si *SSMInstaller) Run(ctx context.Context, req SSMRunRequest) error {
+		Status:     status,
+	}
+}
+
+/*
+SSM SendCommand failed with ErrCodeInvalidInstanceId. Make sure that the instances have AmazonSSMManagedInstanceCore policy assigned.
+Also check that SSM agent is running and registered with the SSM endpoint on that instance and try restarting or reinstalling it in case of issues.
+See https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_SendCommand.html#API_SendCommand_Errors for more details.
+*/
+
+func (si *SSMInstaller) emitInvalidInstanceEvents(ctx context.Context, req SSMRunRequest, instancesState *instancesSSMState) error {
+	for _, instanceID := range instancesState.missing {
+		event := invalidSSMInstanceEvent(req.AccountID, req.Region, instanceID,
+			"EC2 Instance is not registered in SSM. Make sure that the instance has AmazonSSMManagedInstanceCore policy assigned.",
+		)
+		if err := si.Emitter.EmitAuditEvent(ctx, &event); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	for _, instanceID := range instancesState.connectionLost {
+		event := invalidSSMInstanceEvent(req.AccountID, req.Region, instanceID,
+			"SSM Agent in EC2 Instance is not connecting to SSM Service. Restart or reinstall the SSM service. See https://docs.aws.amazon.com/systems-manager/latest/userguide/ami-preinstalled-agent.html#verify-ssm-agent-status for more details.",
+		)
+		if err := si.Emitter.EmitAuditEvent(ctx, &event); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	for _, instanceID := range instancesState.unsupportedOS {
+		event := invalidSSMInstanceEvent(req.AccountID, req.Region, instanceID,
+			"EC2 instance is running an unsupported Operating System. Only Linux is supported.",
+		)
+		if err := si.Emitter.EmitAuditEvent(ctx, &event); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+type instancesSSMState struct {
+	valid          []string
+	missing        []string
+	connectionLost []string
+	unsupportedOS  []string
+}
+
+// ssmAgentState returns the instancesSSMState for all the instances.
+func (si *SSMInstaller) ssmAgentState(ctx context.Context, req SSMRunRequest, allInstanceIDs []string) (*instancesSSMState, error) {
+	ret := &instancesSSMState{}
+
+	// Default is 10, but AWS returns an error if less than 5.
+	maxResults := aws.Int64(int64(10))
+	if len(allInstanceIDs) > 10 {
+		maxResults = aws.Int64(int64(len(allInstanceIDs)))
+	}
+
+	ssmInstancesInfo, err := req.SSM.DescribeInstanceInformationWithContext(ctx, &ssm.DescribeInstanceInformationInput{
+		Filters: []*ssm.InstanceInformationStringFilter{
+			{Key: aws.String(ssm.InstanceInformationFilterKeyInstanceIds), Values: aws.StringSlice(allInstanceIDs)},
+		},
+		MaxResults: maxResults,
+	})
+	if err != nil {
+		// Ignore AccessDeniedException error because users might not have granted `ssm:DescribeInstanceInformation` policy.
+		// Previous docs didn't require this Policy.
+		awsErr := awslib.ConvertRequestFailureError(err)
+		if trace.IsAccessDenied(awsErr) {
+			return nil, trace.Wrap(awsErr)
```

Comment:
AccessDenied is handled the same as any other error, no?

Reply:
Yes 👍 removed